Add directory creation method that takes an ACL

carlossanlop opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/41614

@JeremyKuhne

carlossanlop commented on src/Common/src/CoreLib/Interop/Windows/Kernel32/Interop.SECURITY_ATTRIBUTES.cs:
```diff
@@ -15,6 +15,7 @@ internal struct SECURITY_ATTRIBUTES
             internal uint nLength;
             internal IntPtr lpSecurityDescriptor;
             internal BOOL bInheritHandle;
+            internal unsafe byte* pSecurityDescriptor;
```
Brought this from framework. The only difference is that it does not assign null by default. Should I?

jkotas replied:
This structure is defined by the Windows SDK. The [documentation](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/aa379560(v=vs.85)) does not say anything about `pSecurityDescriptor` field that you are adding. What is this field for?

JeremyKuhne replied:
Yeah, don't add this. Assign to `lpSecurityDescriptor` as I've described below.

carlossanlop replied:
Thanks for noticing. Yes, I'll remove it.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,75 @@ namespace System.IO
+        /// ## Remarks
+        ///
+        /// This method can also throw any of the exceptions that <xref:System.IO.FileStream> can throw.
+        /// ]]></format></remarks>
```
Can't have a new public API without documentation 😃

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,75 @@ namespace System.IO
+        /// ]]></format></remarks>
+        public static FileStream Create(this FileInfo fileInfo, FileMode mode, FileSystemRights rights, FileShare share, int bufferSize, FileOptions options, FileSecurity fileSecurity)
+        {
+            string badArg = null;
```
Based my checks on what we already do inside the FileStream constructor that receives all these arguments.
Question: Should I delegate the argument checks and exception throws to the other public APIs I will be calling, like the FileStream constructor, or should I also check them here?

JeremyKuhne replied:
You want throwing to occur in the same order as much as possible, so I'm guessing you'll have to duplicate the checks.

carlossanlop replied:
Ok thanks for confirming. I don't mind duplicating checks. I wanted to make sure I also had to throw here. I'll make sure they are thrown in the same order whenever possible.

carlossanlop replied:
This will be addressed in the File Create PR.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,75 @@ namespace System.IO
+
+            // I cannot do this (which is what FileSystem.CreateDirectory() currently does)
+            // unless we want a circular dependency between System.IO.FileSystem and System.IO.FileSystem.AccessControl
+
```
@JeremyKuhne this is what's causing me conflict: I am already consuming System.IO.FileSystem.AccessControl as a new dependency for System.IO.FileSystem, so I cannot call FileSystem.CreateDirectory here unless I have a circular dependency. The other problem I have is that I cannot invalidate the directoryInfo here because Invalidate() is private.

carlossanlop replied:
Update: I think I will have to do something like in [this PR](https://github.com/dotnet/corefx/pull/26395/files), where the `System.IO.Pipes.AccessControl` APIs are declared in the ref files, but the source code is written in `System.IO.Pipes`.

I noticed this could be the solution after I submitted [my other PR](https://github.com/dotnet/corefx/pull/41838) for the implementation of the creation APIs for Pipes.

JeremyKuhne replied:
Just include the sources. I gave a sample on how you should change the signature.

carlossanlop replied:
Ok I think I have covered this part, all the sources were included (including the dependencies of these dependencies). I only need to figure out one error, and I'll push a new commit.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/Security/AccessControl/FileSecurity.cs:
```diff
@@ -17,7 +17,6 @@ public FileSecurity()
         public FileSecurity(string fileName, AccessControlSections includeSections)
             : base(false, fileName, includeSections, false)
         {
-            Path.GetFullPath(fileName);
```
This call was being done inside the constructor. I removed it but I wonder if it was added here to throw in case the full path can't be obtained.

JeremyKuhne replied:
Yes, do not remove it. Please add a comment that we're depending on GetFullPath validation.

carlossanlop commented on src/System.IO.FileSystem/src/System.IO.FileSystem.csproj:
```diff
@@ -326,4 +326,8 @@
   <ItemGroup Condition="'$(TargetsUnix)' == 'true'">
     <Reference Include="System.Threading" />
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetsWindows)' == 'true'">
```
Is this the correct condition, or should I add more? See the top of this same file, there are many windows condition combinations.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:
```diff
@@ -2,8 +2,6 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
```
Here and below, these usings were unused.

JeremyKuhne replied:
They're used by the redist build.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -111,6 +109,24 @@ public static void CreateDirectory(string fullPath)
+                    fixed (byte* sdPtr = sd)
+                    {
+                        byte* bytesOnStack = stackalloc byte[sd.Length];
+                        Buffer.MemoryCopy(sdPtr, bytesOnStack, sd.Length, sd.Length);
```
Framework had MemCpy, Core has MemoryCopy, with a different parameter order. Was this the correct method to use?

JeremyKuhne replied:
Neither. Pin it.

JeremyKuhne replied:
The Framework code isn't right. Copying to the stack is unnecessary- it could have just been pinned there as well.

carlossanlop commented on src/System.Security.AccessControl/src/System/Security/AccessControl/NativeObjectSecurity.cs:
```diff
@@ -281,9 +281,7 @@ private void Persist(string name, SafeHandle handle, AccessControlSections inclu
-                            exception = new ArgumentException(
-                                 SR.Argument_InvalidName,
-nameof(name));
+                            exception = new ArgumentException(SR.Argument_InvalidName, nameof(name));
```
Small spacing fix.

danmosemsft commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +186,27 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            FileInfo info = new FileInfo(path);
+            FileSecurity security = new FileSecurity();
+            using FileStream stream = info.Create(FileMode.CreateNew, FileSystemRights.FullControl, FileShare.ReadWrite, 1, FileOptions.None, security);
+            Assert.True(File.Exists(path));
```
Please read back the ACL's . It would be not good to set the wrong ones or fail to set them :)

carlossanlop replied:
Good point. That's the whole reason of this PR. 😄

JeremyKuhne replied:
A good choice would be to give full rights to "Users" as you'll be able to validate it reasonably easily.

danmosemsft replied:
Note that, if we get this wrong (eg set the wrong ACL) then servicing our code won't help. The ACL's will already have been written to filesystems everywhere. It will be quite bad.

carlossanlop replied:
@danmosemsft I added verifications in the unit tests in case you want to take another look :)

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -59,7 +57,7 @@ public static void ReplaceFile(string sourceFullPath, string destFullPath, strin
         }
 
-        public static void CreateDirectory(string fullPath)
+        public static void CreateDirectory(string fullPath, DirectorySecurity dirSecurity = null)
```
Have this take the security descriptor instead, then there will be no dependencies.

``` C#
namespace System.IO
{
    public static class FileSystemAclExtensions
    {
        public static void Create(
            this DirectoryInfo directoryInfo,
            DirectorySecurity directorySecurity)
        {
            // Parameter validation here

            FileSystem.CreateDirectory(directoryInfo.FullName, directorySecurity.GetSecurityDescriptorBinaryForm());
        }
    }

    internal static partial class FileSystem
    {
        public unsafe static void CreateDirectory(string fullPath, byte[] securityDescriptor = null)
        {
            // Existing goo

            fixed (byte* b = securityDescriptor)
            {
                var securityAttributes = new Interop.Kernel32.SECURITY_ATTRIBUTES
                {
                    nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES),
                    lpSecurityDescriptor = (IntPtr)b
                };

                // Existing create directory loop here, eg
                // while()
                // {
                //  r = Interop.Kernel32.CreateDirectory(name, ref secAttrs);
                // }
            }
        }
    }
}
```

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -111,6 +109,24 @@ public static void CreateDirectory(string fullPath)
 
+            if (dirSecurity != null)
+            {
+                secAttrs.nLength = (uint)Marshal.SizeOf(secAttrs);
```
Don't ever use `Marshal.SizeOf`. You should only be using it if our struct is not blittable, which would be a problem.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,69 @@ namespace System.IO
+                throw new ArgumentNullException(nameof(directorySecurity));
+
+            // FileSystem is an internal class in another assembly. How do I consume it here?
+            // FileSystem.CreateDirectory(directoryInfo.FullName, directorySecurity.GetSecurityDescriptorBinaryForm());
```
I tried adding FileSystem.Windows.cs directly in FileSystem.AccessControl.csproj, but that file has a lot of additional dependencies. I wouldn't mind adding them if they were all in Common, but it also depends on other internal classes like PathInternal. So I'm not sure if I should continue pursuing that option, or if there's a better way.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,69 @@ namespace System.IO
+                throw new ArgumentNullException(nameof(fileSecurity));
+
+            fileInfo.SetAccessControl(fileSecurity);
+            return new FileStream(fileInfo.FullName, mode, mode == FileMode.Append ? FileAccess.Write : FileAccess.ReadWrite, share, bufferSize, options);
```
@JeremyKuhne I have some questions about these two lines above:

1. Is calling SetAccessControl (defined in this same file) the correct way to set the access control for the fileSecurity object, before the file actually exists? It will internally call the Persists method, which doesn't seem to care if the file doesn't exist, because it knows it should pass a null SafeHandle in that case ([source](https://github.com/dotnet/corefx/blob/0cc22ccba707d7552159363d586262fa9c4e8a2a/src/System.Security.AccessControl/src/System/Security/AccessControl/NativeObjectSecurity.cs#L349)).

2. The FileAccess value I'm passing to the FileStream constructor is based on the FileMode value that was passed to the Create method. Is that the correct thing to do? We do that in one of the FileStream constructors ([source](https://github.com/dotnet/corefx/blob/ac99a1b7168bd32046a954c3f06012c0fa909bed/src/Common/src/CoreLib/System/IO/FileStream.cs#L166)).

carlossanlop replied:
This will be addressed in the File Create PR.

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -107,9 +107,25 @@ public static void CreateDirectory(string fullPath)
+            {
+                unsafe
+                {
+                    fixed (byte* pSecurityDescriptor = securityDescriptor)
```
This will fail. As soon as you leave the fixed statement the byte array will unpin. If the garbage collector runs the array will move and you'll get an AV or worse.

All of the logic that accesses SECURITY_ATTRIBUTES must be in the fixed block.

Additionally you do not need to check if the passed in array is null. Just pin it and you'll get a null pointer, which is what you want anyway.

JeremyKuhne replied:
Note that it was technically incorrect to not set the size before.

carlossanlop replied:
Ok I'll fix it.

carlossanlop committed:
First commit

carlossanlop committed:
Fix OperationCanceledException dependency error, revert nullable indication

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    internal static partial class FileSystem
+    {
+        internal static string GetDirectoryRoot(string path)
```
@JeremyKuhne this internal method was originally inside the Directory class, but it was only used by FileSystem, once in Windows, once in Unix. I was having issues splitting Directory and only importing the partial file with this method, because FileSystem.AccessControl was already consuming the assembly where Directory was published, but I was trying to force compiling an extra internal section of that same class. I fixed it by moving this method to FileSystem, in a new file shared by all architectures.

JeremyKuhne replied:
That's fine.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.AccessControl.Windows.cs:
```diff
@@ -0,0 +1,249 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
@JeremyKuhne do you agree with the name for this file?

JeremyKuhne replied:
I'd make two: `FileSystem.DirectoryCreation.Windows.cs` and `FileSystem.Attributes.Windows.cs`.  That would be much more understandable, imho.

carlossanlop replied:
Will do

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/Directory.cs:
```diff
@@ -232,12 +232,6 @@ public static string GetDirectoryRoot(string path)
             return root;
         }
 
-        internal static string InternalGetDirectoryRoot(string path)
```
Moved to a new file called FileSystem.cs. See full explanation of this change in the comment I left in that file.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/Resources/Strings.resx:
```diff
@@ -143,4 +143,175 @@
     <value>Remove access rule failed</value>
   </data>
-</root>
+  <data name="ArgumentNull_Buffer" xml:space="preserve">
```
These are all resource strings consumed by the newly added files coming from IO.FileSystem.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System.IO.FileSystem.AccessControl.csproj:
```diff
@@ -17,6 +17,138 @@
     <Compile Include="System\Security\AccessControl\FileSystemRights.cs" />
     <Compile Include="System\Security\AccessControl\FileSystemSecurity.cs" />
     <Compile Include="System\IO\FileSystemAclExtensions.cs" />
+    <Compile Include="..\..\System.IO.FileSystem\src\Microsoft\Win32\SafeHandles\SafeFindHandle.Windows.cs">
```
These are all dependencies from the files that we needed from IO.FileSystem.

stephentoub replied:
Should we move the files to common instead?

carlossanlop replied:
If that's the correct procedure, then sure, I'll do it. Are there any cases where the file shouldn't be moved to Common?

carlossanlop replied:
I moved them. I'd still like to know if there are exceptions to the rule of moving shared files to common.

carlossanlop committed:
Only Directory files

carlossanlop commented:
Due to the size of this change, his PR will only include the changes for the Directory Create method. I'll submit the File Create method in a new PR.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Windows.cs:
```diff
@@ -59,110 +59,6 @@ public static void ReplaceFile(string sourceFullPath, string destFullPath, strin
             }
         }
 
-        public static void CreateDirectory(string fullPath)
```
Moved to the new file FileSystem.AccessControl.Windows.cs

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.AccessControl.Windows.cs:
```diff
@@ -0,0 +1,249 @@
+            {
+                unsafe
+                {
+                    fixed (byte* pSecurityDescriptor = securityDescriptor)
```
This needs to be fixed per my earlier comments. There is no need to check for null and the loop below needs to be in the fixed block.

carlossanlop replied:
Yes, sorry, I fixed it now.

carlossanlop committed:
Unit tests verifications

carlossanlop commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +187,48 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
                 FileSystemAclExtensions.SetAccessControl(fileStream, fileSecurity);
             }
         }
+
```
@JeremyKuhne I added one more test. I am not sure what else I should verify. Can you please take a look and tell me what you think?

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -188,14 +189,46 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            CreateAndVerifyDirectoryWithSecurity(security);
+        }
+        [Theory]
+        [InlineData("S-1-5-32-544", FileSystemRights.FullControl, AccessControlType.Allow)]
```
I would not use `Administrators` as you will potentially block tests when run by accounts that are not in the group. Use [`WellKnownSidType`] to be clearer, and give the rights to `BuiltinUsersSid`, which should be the `Users` group. You can validate what is what by getting/setting permissions in the UI and getting/setting the ACL with the current extension methods.

carlossanlop replied:
Ok I'll use that.

carlossanlop replied:
Done.

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -188,14 +189,46 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            Assert.Equal(typeof(FileSystemRights), security.AccessRightType);
+
+            DirectorySecurity retrievedSecurity = info.GetAccessControl();
+            Assert.Equal(typeof(FileSystemRights), retrievedSecurity.AccessRightType);
```
Why aren't you actually verifying the rights on the object?

carlossanlop replied:
I was investigating what exactly to compare. I pushed another commit. Can you please review the changes?

carlossanlop replied:
I just saw your other comment about using WellKnownSidType. I'll address that.

carlossanlop replied:
Fixed.

carlossanlop committed:
Split FileSystem.Windows into two additional files instead of one. Fix pinning.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:
```diff
@@ -0,0 +1,132 @@
+            int firstError = 0;
+            string errorString = fullPath;
+
+            unsafe
```
@JeremyKuhne removed the null check, made sure to put the while inside the fixed section.
Also split the FileSystem file into the two you suggested, and added two unit tests.

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:
```diff
@@ -0,0 +1,132 @@
+            int firstError = 0;
+            string errorString = fullPath;
+
+            unsafe
```
nit: we typically just make the method unsafe, it helps reduce the nesting

carlossanlop replied:
I could do that, although I read the other day that it's recommended to wrap the smallest amount of code with the unsafe section.

JeremyKuhne replied:
It is one approach, but we don't do that. I've been explicitly asked by our architects to make the methods themselves unsafe.

carlossanlop replied:
Okay no problem.

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:
```diff
@@ -0,0 +1,132 @@
+                        lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                    };
+
+                    // If all the security checks succeeded create all the directories
```
nit- this comment is dead- can you remove it?

carlossanlop replied:
Thanks for noticing. I removed it.

carlossanlop committed:
Remove dead comment, add actual UT verifications

carlossanlop committed:
User WellKnownSidType

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,79 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            Assert.Throws<ArgumentNullException>(() => info.Create(null));
+        }
+
+
```
nit: extra space

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,79 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            DirectorySecurity security = new DirectorySecurity();
+            CreateAndVerifyDirectoryWithSecurity(security);
+        }
+        [Theory]
```
nit: missing line break

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,79 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+        private bool AreRulesEqual(FileSystemAccessRule expectedRule, FileSystemAccessRule actualRule)
+        {
+            return
+            expectedRule.AccessControlType == actualRule.AccessControlType &&
```
nit: indent

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,79 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            expectedRule.FileSystemRights  == actualRule.FileSystemRights &&
+            expectedRule.InheritanceFlags  == actualRule.InheritanceFlags &&
+            expectedRule.PropagationFlags  == actualRule.PropagationFlags;
+            
```
nit extra line

JeremyKuhne commented on src/System.IO.FileSystem/src/System.IO.FileSystem.csproj:
```diff
@@ -20,6 +20,8 @@
     <Compile Include="System\IO\Enumeration\FileSystemEnumerableFactory.cs" />
     <Compile Include="System\IO\Enumeration\FileSystemEnumerator.cs" />
     <Compile Include="System\IO\Enumeration\FileSystemName.cs" />
+    <Compile Include="System\IO\FileSystem.DirectoryCreation.Windows.cs" />
```
This is in the wrong section, it needs to be in the item group that is conditioned to windows.

carlossanlop replied:
Oooh nice catch. I used VS to split it into Attributes + DirectoryCreation. Didn't notice the newly duplicated file got added to the wrong section.

carlossanlop committed:
Make method unsafe, nit comments addressed.

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,19 @@ namespace System.IO
+            if (directorySecurity == null)
+                throw new ArgumentNullException(nameof(directorySecurity));
+
+            // FileSystem is an internal class in another assembly. How do I consume it here?
```
Comment no longer relevant

carlossanlop replied:
```suggestion
```

carlossanlop committed:
Remove irrelevant comment

carlossanlop commented:
I addressed the latest comments, @JeremyKuhne. I just need to make sure I indicate all the expected exceptions in the triple slash comments, and that I also have some unit tests to verify those exceptions.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,18 @@ namespace System.IO
+        /// <exception cref="ArgumentNullException"><paramref name="directorySecurity" /> is <see langword="null" />.</exception>
+        public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity directorySecurity)
+        {
+            if (directorySecurity == null)
```
@JeremyKuhne should we make this `if (string.IsNullOrWhiteSpace(directorySecurity))` instead? Or do we let `FileSystem.CreateDirectory` handle the empty/whitespace case?

JeremyKuhne replied:
Not sure what you mean. `directorySecurity` isn't a string.

carlossanlop replied:
Oops my bad. I wrote this before my morning coffee.

carlossanlop replied:
Actually, I remember why I wrote this _apparently dumb_ question: It was directed towards the `directoryInfo.FullName` argument, which we pass to the CreateDirectory method, but we don't check if it's NullOrWhiteSpace, we just check if the directory exists.

JeremyKuhne replied:
Yeah, I think we're good leaving that to the DirectoryInfo construction

carlossanlop committed:
Add additional exceptions to triple slash

carlossanlop committed:
Add unit tests for additional exceptions

JeremyKuhne commented on src/System.IO.FileSystem/src/System/IO/FileSystem.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    internal static partial class FileSystem
+    {
+        internal static string GetDirectoryRoot(string path)
```
Looking into this a bit further, this is a relic of the additional overhead we used to have. You should just remove this file and call `Path.GetPathRoot()` directly.

carlossanlop committed:
Remove FileSystem.cs, it has one obsolete method. Replace all calls of that removed method.

carlossanlop commented on src/System.IO.FileSystem.AccessControl/src/System.IO.FileSystem.AccessControl.csproj:
```diff
@@ -6,20 +6,152 @@
     <Compile Include="System\Security\AccessControl\FileSystemAuditRule.cs" />
     <Compile Include="System\Security\AccessControl\FileSystemRights.cs" />
     <Compile Include="System\Security\AccessControl\FileSystemSecurity.cs" />
-    <Compile Include="System\IO\FileSystemAclExtensions.cs" />
```
I sorted these `<Compile>` items inside the Windows item group alphabetically. It was driving me crazy trying to find them when I was adding the bunch of dependencies from System.IO.FileSystem.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:
```diff
@@ -0,0 +1,129 @@
+{
+    internal static partial class FileSystem
+    {
+        public static unsafe void CreateDirectory(string fullPath, byte[] securityDescriptor = null)
```
@JeremyKuhne the `string fullPath` is where I wanted to ask you about `string.IsNullOrWhitespace`, not the `DirectorySecurity` object. We don't check it. Should we?

JeremyKuhne replied:
DirectoryInfo does.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/Directory.cs:
```diff
@@ -227,17 +227,11 @@ public static string GetDirectoryRoot(string path)
 
             string fullPath = Path.GetFullPath(path);
-            string root = fullPath.Substring(0, PathInternal.GetRootLength(fullPath.AsSpan()));
+            string root = Path.GetPathRoot(fullPath);
```
@JeremyKuhne found this 3rd instance where GetPathRoot could be called.

JeremyKuhne replied:
This one I wouldn't change as this does add measurable overhead to a simple API.

carlossanlop commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -286,7 +286,7 @@ public static void CreateDirectory(string fullPath)
             if (count == 0 && !somepathexists)
             {
-                string root = Directory.InternalGetDirectoryRoot(fullPath);
+                string root = Path.GetPathRoot(fullPath);
```
This is the 2nd instance where GetPathRoot can be called.

JeremyKuhne reviewed (approved):
Looks good

carlossanlop commented:
@JeremyKuhne thanks.

The CI build is failing with this error:

```
.packages\microsoft.dotnet.apicompat\5.0.0-beta.19517.8\build\Microsoft.DotNet.ApiCompat.targets#L82
.packages\microsoft.dotnet.apicompat\5.0.0-beta.19517.8\build\Microsoft.DotNet.ApiCompat.targets(82,5): error : MembersMustExist :
Member 'System.IO.FileSystemAclExtensions.Create(System.IO.DirectoryInfo, System.Security.AccessControl.DirectorySecurity)' does not exist in the implementation but it does exist in the contract.
```

Do you know what did I miss?

stephentoub commented on src/System.IO.FileSystem.AccessControl/src/Resources/Strings.resx:
```diff
@@ -143,4 +143,175 @@
+  <data name="OperationCanceled" xml:space="preserve">
+    <value>The operation was cancelled.</value>
+  </data>
+</root>
```
All of these resources are required?

carlossanlop replied:
I could have to check again. When I added them, it was because of the dependency that I included to FileSystem.CreateDirectory, which calls code from PathInternal, and needed these resources. Then I trimmed the code but I didn't check these resources. I'll do another pass.

carlossanlop replied:
I removed a few.

stephentoub commented on src/System.IO.FileSystem.AccessControl/src/System.IO.FileSystem.AccessControl.csproj:
```diff
@@ -6,20 +6,152 @@
+      <Link>Common\System\Memory\FixedBufferExtensions.cs</Link>
+    </Compile>
+    <Compile Include="$(CommonPath)\System\NotImplemented.cs">
+      <Link>Common\Interop\Windows\mincore\NotImplemented.cs</Link>
```
This Link path looks wrong.

carlossanlop replied:
I see what you mean, it doesn't look like the rest of the linked files. It was like that already (line 21 from the before file). I can rename it.

stephentoub commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -10,6 +10,20 @@ namespace System.IO
+            if (directorySecurity == null)
+                throw new ArgumentNullException(nameof(directorySecurity));
+
+            FileSystem.CreateDirectory(directoryInfo.FullName, directorySecurity.GetSecurityDescriptorBinaryForm());
```
This will null ref if directoryInfo is null. Don't we usually throw ArgumentNullException for such cases, even for extension methods?

carlossanlop replied:
I didn't know extension methods could be called from a null argument that was passed as `this`. None of the other extension methods in this class do a null check. What would be the correct thing to do then?

carlossanlop replied:
I've been reading about this. I will definitely add the check to throw ArgumentNullException if null.
@stephentoub @JeremyKuhne  Should I add the same check to the existing extension methods?

JeremyKuhne replied:
Yes, but that can happen in 5.0

carlossanlop replied:
I opened an issue to track this: https://github.com/dotnet/corefx/issues/41907

stephentoub replied:
Thanks. I don't actually have a strong preference on these; we should just be consistebt.

@terrajobst, @bartonjs, what do our design guidelines say?

bartonjs replied:
> what do our design guidelines say [for `null` in extension methods]?

Nothing written down.  The problem is (syntactically speaking) from an extension invocation it should be NullReferenceException and from static invocation it should be ArgumentNullException.  Given that we don't know which call style it was, the proposed guidelines I have are

* AVOID defining extension methods that accept a `null` value for a `this` parameter.
* DO NOT allow the `this` parameter being `null` produce a NullReferenceException.

Though, having sat on that for a while, I can now edit it to a much cleaner

* DO throw `ArgumentNullException` when the `this` parameter is `null` in an extension method.

[(null).Something() succeeding confuses people, don't nullref because of static calling conventions, et cetera]

danmosemsft replied:
I don't have a strong preference either, but my assumption is that extension methods are almost always called as "extension invocation" which would suggest NRE would be less confusing. But, my impression is that we are generally using ANE in our own code today.

danmosemsft replied:
Counter argument -- if you are using "static invocation" you might be glad to have the parameter name. Anyway - I say go with whatever @bartonjs writes down 😸

carlossanlop replied:
Thanks everyone.

stephentoub commented on src/System.IO.FileSystem.AccessControl/src/System.IO.FileSystem.AccessControl.csproj:
```diff
@@ -6,20 +6,152 @@
+    <Compile Include="System\IO\FileSystemAclExtensions.cs" />
     <Compile Include="System\Security\AccessControl\DirectoryObjectSecurity.cs" />
     <Compile Include="System\Security\AccessControl\DirectorySecurity.cs" />
+    <Compile Include="System\Security\AccessControl\FileSecurity.cs" />
```
This is a ton of files being added. They're all required?

carlossanlop replied:
Yes. I had to add one by one until I could compile.

carlossanlop replied:
Actually I found a few that were not required. I removed them.

stephentoub commented on src/System.IO.FileSystem/src/System/IO/FileSystem.Attributes.Windows.cs:
```diff
@@ -0,0 +1,128 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Did any code in this file change, or is it just moved without modification from the original source file?

carlossanlop replied:
Just moved, no changes.

stephentoub commented on src/System.IO.FileSystem/src/System/IO/FileSystem.DirectoryCreation.Windows.cs:
```diff
@@ -0,0 +1,129 @@
+                        // time we try using the directory.  Thirdly, it could
+                        // fail because the target does exist, but is a file.
+                        if (currentError != Interop.Errors.ERROR_ALREADY_EXISTS)
+                            firstError = currentError;
```
Nit: as long as you're editing this code, please add braces around the body of this if block

carlossanlop replied:
Sure no problem, I can add the missing braces.

carlossanlop committed:
Second pass on required strings.resx resources

carlossanlop committed:
Remove unused dependency cs files

carlossanlop committed:
Move shared files to Common

carlossanlop committed:
Add the missing braces

carlossanlop committed:
Add nulcheck for this argument, add exception, document exception in triple slash, add unit test.

carlossanlop committed:
Fix netstandard CI build failure

stephentoub commented:
> Are there any cases where the file shouldn't be moved to Common

Generally if a file is used in multiple projects, it's better to put it in common rather than having the file in one project's directory.

danmosemsft commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,102 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+            Assert.Equal(expectedRules.Count, actualRules.Count);
+            if (expectedRules.Count > 0)
+            {
+                Assert.True(actualRules.Count > 0);
```
this line is redundant to 269.

carlossanlop replied:
Ah, thanks. Removed.

danmosemsft commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,102 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+                expectedRule.AccessControlType == actualRule.AccessControlType &&
+                expectedRule.FileSystemRights  == actualRule.FileSystemRights &&
+                expectedRule.InheritanceFlags  == actualRule.InheritanceFlags &&
+                expectedRule.PropagationFlags  == actualRule.PropagationFlags;
```
what about `IsInherited`  and `IdentityReference` properties -- do they need comparing also?

danmosemsft replied:
Looks like `IsInherited` will always be false.

carlossanlop replied:
Let me check.

danmosemsft commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,102 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+
+            List<FileSystemAccessRule> actualRules = actualSecurity.GetAccessRules(includeExplicit: true, includeInherited: false, typeof(SecurityIdentifier))
+                .Cast<FileSystemAccessRule>().ToList();
+
```
What about `GetAuditRules` -  should we also pull those (with `includeInherited: false`) and check there (presumably) aren't any?

danmosemsft replied:
Looks like you would have to do `security.AddAuditRule` in the test as well. I guess at this point we are really testing `DirectorySecurity` and its friends rather than your code.

danmosemsft replied:
Which is not a terribly bad thing, up to you

carlossanlop replied:
I could add those tests too. @JeremyKuhne what do you think? Is it enough with what I have or do we want to be super paranoid?

danmosemsft commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,102 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+        [Fact]
+        public void DirectoryInfo_Create_AccessDenied()
+        {
+            DirectoryInfo info = new DirectoryInfo(@"C:\\Windows\\System32\\deniedaccess");
```
Will this pass if I run the test as admin? Maybe it should bail out if `PlatformDetection.IsWindowsAndElevated` is true.

carlossanlop replied:
I'm not sure. Should I expect these unit tests to always be executed in elevated mode at some point, or should I indicate that condition among the test method attributes? This method in particular should not be run in elevated mode, but I could add one that does, and I would have to investigate the expected result.

danmosemsft replied:
Whatever our official runs do (I assume none are), all our tests should pass whether I run elevated or not. There are some tests that require elevation and are attributed `[ConditionalFact(nameof(IsProcessElevated))]`. This may be an unusual reverse case. Not suggesting you have to verify it but an `if` would do no harm.

danmosemsft commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -187,5 +190,102 @@ public void SetAccessControl_FileStream_FileSecurity_Success()
+        [Theory]
+        [InlineData(WellKnownSidType.BuiltinUsersSid, FileSystemRights.FullControl, AccessControlType.Allow)]
+        [InlineData(WellKnownSidType.BuiltinUsersSid, FileSystemRights.ReadAndExecute, AccessControlType.Allow)]
+        [InlineData(WellKnownSidType.BuiltinUsersSid, FileSystemRights.Write, AccessControlType.Allow)]
```
Maybe throw in a deny just for fun? Although again that's testing existing code.

carlossanlop replied:
Sure, I can add a different combination. I will have to investigate what is the expected result for it.

danmoseley reviewed (approved):
LGTM. I also diffed the moved code.

carlossanlop committed:
Fix Microsoft.IO.Redist build issue with NetStandard

carlossanlop committed:
Remove unnecessary assert in ut

carlossanlop committed:
Fix build problem in Windows Packaging All Configurations x64_Debug, System.Runtime needs to be included to recognize NotNullIfNotNullAttribute, consumed in Path.Internal.cs

carlossanlop commented:
Update:

I made the requested change:  Throw `ArgumentNullException` if `this` is `null`.

I added a unit test that expects `ArgumentNullException` to be thrown when `DirectoryInfo` is `null` and its extension method is invoked.

The unit test passed in all configurations except in NetFX x86 ([helix log](https://helix.dot.net/api/2019-06-17/jobs/ab62ca43-e071-4170-b903-3e682f01a568/workitems/System.IO.FileSystem.AccessControl.Tests/console)). I'm investigating why the output was different in this configuration.

Command used by Helix:
```
eng\common\msbuild.ps1 -warnaserror:0 -ci eng/sendtohelix.proj /t:test /p:ArchGroup=x86 /p:ConfigurationGroup=Release /p:OSGroup=Windows_NT /p:TestScope=innerloop /p:TargetGroup=netfx 
/p:HelixTargetQueues=Windows.10.Amd64.ClientRS5.Open /p:HelixBuild=20191018.45 /p:HelixAccessToken= /p:Creator=dotnet-bot /p:OfficialBuildId= /bl:D:\a\1\s/artifacts/log/Release/SendToHelix.binlog
```

Output:
```
 System.IO.FileSystemAclExtensionsTests.DirectoryInfo_Create_NullDirectoryInfo [FAIL]
      Assert.Throws() Failure
      Expected: typeof(System.ArgumentNullException)
      Actual:   typeof(System.NullReferenceException): Object reference not set to an instance of an object.
      ---- System.NullReferenceException : Object reference not set to an instance of an object.
```

Code:
```cs
[Fact]
public void DirectoryInfo_Create_NullDirectoryInfo()
{
    DirectoryInfo info = null;
    DirectorySecurity security = new DirectorySecurity();
    Assert.Throws<ArgumentNullException>(() => info.Create(security));
}
```

carlossanlop committed:
Additional file splitting and dependency addition to fix netstandard/netfx build problems

carlossanlop committed:
Fix netstandard and netcoreapp build issues

carlossanlop commented:
Update, @JeremyKuhne :
The build problems for netstandard were fixed locally: `build.cmd -allconfigurations` finished successfully. I'm hoping the CI tells us the same. Changes:

- Had to split the `FileSystemAclExtensions.cs` file into one for netcoreapp and one for netstandard, due to the way the dependencies were used.
- I had to do some additional file splitting from path related methods, and reverted a couple that were not necessary anymore.
- I made sure I didn't break `Microsoft.IO.Redist` and `System.IO.FileSystem`, which depend on the modified files.
- Cleaned the csproj files: used `<Compile Include="" Link=""` />, it's easier to read and the files are shorter.

Pending addressing the unit test suggestions by @danmosemsft .

carlossanlop commented on src/Common/src/CoreLib/System/IO/Path.Root.Windows.cs:
```diff
@@ -0,0 +1,26 @@
+    // Provides methods for processing file system strings in a cross-platform manner.
+    // Most of the methods don't do a complete parsing (such as examining a UNC hostname),
+    // but they will handle most string operations.
+    public static partial class Path
```
I should remove this file.

carlossanlop committed:
Remove unused file

carlossanlop committed:
nit cleaning

danmosemsft commented:
Any other feedback @stephentoub ?

stephentoub commented on src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs:
```diff
@@ -2,14 +2,14 @@
-    public static class FileSystemAclExtensions
+    public static partial class FileSystemAclExtensions
     {
+
```
Nit: why was a blank line added here?

carlossanlop replied:
It was a leftover from when I split this file.

stephentoub reviewed (approved):
LGTM once CI is green.

carlossanlop committed:
Minimal unit tests

JeremyKuhne commented on src/System.IO.FileSystem.AccessControl/tests/FileSystemAclExtensionsTests.cs:
```diff
@@ -2,13 +2,28 @@
     {
+        #region Private members
+
+        //private static readonly Lazy<bool> s_isElevated = new Lazy<bool>(AdminHelpers.IsProcessElevated);
```
Don't leave the commented out code in here. Create an issue to add access denied tests for directories to track.

carlossanlop committed:
Remove unused unit test code and spacing fix

JeremyKuhne reviewed (approved):
Looks good, modulo removing the commented-out code.

carlossanlop committed:
Add full framework condition to unit test

carlossanlop committed:
Additional fix for full framework
